### PR TITLE
UWP initial support

### PR DESCRIPTION
**Pull request by hrydgard**
This adds a new .sln and set of projects that builds PPSSPP for the Universal Windows 10 Platform (UWP).

Performance is quite good on desktop, but worse than expected on mobile Windows 10, at least on my Lumia 650 - it's really quite terrible. What makes it even worse is that Microsoft did add support for code generation (VirtualProtectForApp) but forgot to make FlushInstructionCache available. That's not a big deal on x86 where it's not needed, but on ARM, it's, well, needed. Currently it "works" anyway by logging every time we write code (this is horrible!), which for some reason is enough to flush the instruction cache, and only when running under the debugger so that the logging actually does something. But as it is now, the Windows Mobile 10 build is not really viable on ARM platforms.

I/O outside of your app's private folders is not allowed on UWP except through a nasty async interface to files that you let the user pick using a file chooser. The previous attempt simply copied the ISO into the private folders, but I went ahead and added a terrible wrapper that runs async I/O on a thread, simulating sync I/O, which lets you play ISO files from anywhere by clicking Load... and picking an ISO file. But the integrated directory browser is not yet functional. We do keep the memstick in the app's private folder so no trickery needed there. 

Audio is currently done through XAudio2 (code from @perneky's attempt) but WASAPI is supposed to be available too, need to try that.

### CODE REVIEW

**Comment by perneky** [comment]
Nice work. For flushing the instruction cache. It is possible with just two ASM instructions, you can flush the entire cache, or just specific lines. As for: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0201d/ch03s02s03.html

So a small function could be written in ASM or maybe inline just by using intrinsic functions.

**Comment by hrydgard** [comment]
@perneky , you'd think so, but those instructions are not available in user mode.

**Comment by perneky** [comment]
:( Sad enough.

**Comment by hrydgard** [comment]
The only alternative I can see that might work is to make sure to only write code to new cachelines (by spacing out new blocks to the closes 64-byte boundary, and turn off everything that needs patching of old code like block linking) and make the JIT code block so large that it never needs clearing during a run.

By some lookahead and compiling multiple blocks at once we could eliminate much of that alignment spacing overhead but we'd also risk compiling blocks that will never be accessed.

**Comment by perneky** [comment]
It is really nice that they even provided an intrinsic function for that, so on Win10Mobile this compiles:

static void FlushInstructionCache()
{
#ifdef _M_ARM
  _MoveToCoprocessor( 0, 15, 0, 7, 5, 0 );
#endif // _M_ARM
}

to:

011CD0A0  push        {r4,r5,r11,lr}  
011CD0A4  add         r11,sp,#8  
011CD0A8  movs        r3,#0  
011CD0AA  mcr         p15,#0,r3,c7,c5  
011CD0AE  pop         {r4,r5,r11,pc} 

But throws an illegal instruction exception on mcr. Really nice job. Why are they exposing it if you shouldn't use it?

**Comment by hrydgard** [comment]
They might use that wrapper in the kernel?

Also there are some parameters to mcr that are allowed in user-mode, but cache flushing is not one of them. That's not MS' fault, it's just the way it is on ARM 32-bit, but it is MS' fault that they forgot to expose FlushInstructionCache! So silly.

**Comment by perneky** [comment]
Tried to compile this after merging. When cloned, and built clean for UWP, the git-version.cpp is not generated, so no compile. Managed to work that around by generating it manually from the normal windows build.

It compiled, but upon deployment it can't deploy the compiled package because of some problem in the manifest.

**Comment by hrydgard** [comment]
Oh, right, yeah the git-version thing needs to be fixed.

You should be able to build and deploy using the debugger. Remember to set target to "Device" instead of "Remote". I'm looking into the signing stuff now but it should only be necessary for the official builds that go on the store, right?

**Comment by perneky** [comment]
Right, that is only needed for the store.

I have tried x64, with Local Machine as target. Not tried the phone yet, maybe later this day.

**Comment by hrydgard** [comment]
For me, x64 local machine works just fine.

**Comment by DarkLegend481** [comment]
Hey @hrydgard, I was talking with Xbox Support... Emulators can stay on Xbox Store, but, the emulator needs a license to use games (PPSSPP = Sony license for PSP games). After this, you need a license to put the emulator on Xbox Store... Congratulations!! PPSSPP on Xbox Store!

**Comment by perneky** [comment]
On Win64, when I try to deploy the application to Local Machine, I get:

1>------ Deploy started: Project: PPSSPP_UWP, Configuration: Debug x64 ------
1>Creating a new clean layout...
1>Copying files: Total 50 mb to layout...
1>Checking whether required frameworks are installed...
1>Registering the application to run from layout...
1>DEP0700 : Registration of the app failed. error 0x80070003: Opening file from location: AppxManifest.xml failed with error: The system can not find the path specified.
1>. (0x80073cf0)
========== Deploy: 0 succeeded, 1 failed, 0 skipped ==========

Tried it on ARM with my Lumia950XL. It deploys nicely and the emulator starts. As soon as I try to load Loco Roco, the emulator controls show up, then a crash happens on the UWPRenderThread, with no call stack. From the disassembly, it seems JIT code to me. :(

**Comment by hrydgard** [comment]
Hm, there's a pre-build step that should have copied the correct AppxManifest into place, depending on the configuration (actually, I'm probably gonna get rid of that...)..

If you deploy with just "F5" on the Lumia, so it runs with the debugger attached, it should go a bit further I think. Though I haven't tested Loco. That's definitely a JIT crash, yeah :(  So crazy of MS to forget FlushInstructionCache.

**Comment by hrydgard** [comment]
@DarkLegend481  Yeah, I don't think Sony is going to give me such a license...

**Comment by DarkLegend481** [comment]
@hrydgard PSP it's discontinued gaming platform... Maybe, they give... Or not...

File: UI/EmuScreen.cpp [hunk 1]
```diff
@@ -77,6 +77,9 @@
 #if !PPSSPP_PLATFORM(UWP)
 #include "gfx/gl_common.h"
 #endif
+#if !PPSSPP_PLATFORM(UWP)
+#include "gfx/gl_common.h"
+#endif
```
Comment (unknownbrackets) [inline UI/EmuScreen.cpp]: Oops, this was added twice, likely during a rebase.

-[Unknown]

**Comment by DarkLegend481** [comment]
@hrydgard But good... Even if it is for a few days... Will you make PPSSPP avaliable for Xbox One in the Windows Store?

**Pull request merged**